requirements.txt

```diff
@@ -4,11 +4,11 @@
 
 # Avoid method=restore bug: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4209#issuecomment-1739644328
 # dipy 1.8.x causes conflicts, see https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4319#issuecomment-1864670675
-dipy!=1.6.*,!=1.7.*, !=1.8.*
```

Comment:
Note that when installing `dipy==1.8.0`, we can revisit this older PR of mine as well: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4314#issuecomment-1841570140

Reply:
Hrm... I've done what I think is correct in both https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332/commits/c13446b5a54f077c001d279446bcbf4138bf22e6 and https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332/commits/4db65f8f9e53d9675a92ed3aabdb5a416d69de81, but now our test suite is stalling specifically on one test, specifically on the macOS runners. The last log output we see is...

```
 testing/cli/test_cli_sct_register_multimodal.py::test_sct_register_multimodal_mt0_image_data_within_threshold[False-step=1,algo=slicereg,type=im,iter=5,smooth=0,metric=MeanSquares-mt/mt0_reg_slicereg_goldstandard.nii.gz] PASSED [ 91%]
Error: The operation was canceled
```

The test that comes immediately afterwards is:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8248383221/job/22558595898#step:5:431

Ah! It's `algo=dl`. So... some change I've implemented in this PR is incompatible with `algo=dl` on macOS 12 and 14 runners, but not macOS 11 runners?

I'm going to temporarily add a skip for the `algo=dl` test just to see if my theory is right.

Reply:
Looks like the tests are stalling (or, maybe not stalling, but taking a *very long time*) at one additional point: 

12:

```
testing/cli/test_sct_deepseg.py::test_install_model PASSED               [ 96%]
testing/cli/test_sct_deepseg.py::test_model_dict PASSED                  [ 97%]
```

14:

```
testing/cli/test_sct_deepseg.py::test_install_model PASSED               [ 96%]
testing/cli/test_sct_deepseg.py::test_model_dict PASSED                  [ 97%]
testing/cli/test_sct_deepseg.py::test_segment_nifti[/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2s/t2s.nii.gz-/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2s/t2s_seg-deepseg.nii.gz-t2s_seg_deepseg.nii.gz-seg_sc_t2star-0.9] PASSED [ 97%]
```

`algo=dl` and `sct_deepseg` both share `torch` as a common denominator, so I'll try to dig a little deeper into interactions between `torch`, `numpy`, and the macOS runners.

Reply:
In https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332/commits/80419804947250978aae48f8ebb98924b956d71f, I think I confirmed that the stalls happen for both `macos-11` and `macos-12`. Meaning, I don't _think_ the updates from the `master` branch were the cause of the stalls.

Instead, I'm suspicious of the package updates that naturally occurred between the last passing run (Jan 12) and the first failing run (Mar 15). Given that `torch` seems to be the culprit here, I'm going to start by limiting the versions to see if that makes a difference.

Reply:
If my next set of tests don't work, then I think I might go nuclear and rebase this whole branch onto `master`, then replay each of my old commits one by one until something breaks.

Reply:
Aha! The tests [passed on macos-12](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8288556120/job/22686674485?pr=4332) after I downgraded from `dipy==1.9.0` to `dipy==1.8.0`!

I don't... have any clue why `dipy==1.9.0` would interact with the `torch` tests? But, I'll try to unpin all of the testing changes I made to see if I can isolate the changes.

----

Notably, though the macos-11 test still failed? And, it failed at the same test that was hanging on previous runs:

```
 testing/cli/test_sct_deepseg.py::test_segment_nifti[/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2/t2.nii.gz-/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2/t2_seg-manual.nii.gz-t2_seg_deepseg.nii.gz-seg_sc_contrast_agnostic-None] FAILED [ 97%]

 FAILED testing/cli/test_sct_deepseg.py::test_segment_nifti[/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2/t2.nii.gz-/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/data/sct_testing_data/t2/t2_seg-manual.nii.gz-t2_seg_deepseg.nii.gz-seg_sc_contrast_agnostic-None] - Exception: ('Download error', [ChunkedEncodingError(ProtocolError("Connection broken: TimeoutError(60, 'Operation timed out')", TimeoutError(60, 'Operation timed out')))])
```

A timeout error on downloading the contrast agnostic model? That would explain the apparent stalling, and it would perhaps explain intermittent behavior (i.e. maybe the macos-12 passing test was a fluke?).

Reply:
Ah, retrying it resulted in the `macos-11` tests passing. So, I guess it is a `dipy==1.9.0` issue, and that one timeout was just a really weird coincidence?

Reply:
tl;dr of the minimized comments:

- `dipy==1.8.0` works fine, as does upgrading `numpy` to the newest version.
- However, in the span of Jan-Mar, `dipy==1.9.0` was release as well. And, this version causes our macOS runners to hang on specific `torch` tests for reasons I have yet to identify.

Reply:
I've reported this upstream here:

- https://github.com/dipy/dipy/issues/3125

Just in case the dipy devs have insight into an upstream change that may be causing this behavior.

Reply:
I think for now I'm going to stick with `dipy==1.8.0`, and leave the troubleshooting for `1.9.0` to another PR: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332/commits/1f57611ec42fc2bcc19aaa940659aa5adadab3ed.